services/venice-server/src/main/java/com/linkedin/venice/listener/response/ParallelMultiKeyResponseWrapper.java

```diff
@@ -0,0 +1,133 @@
+   *       that assumption to change, then we have a regression.
+   */
+  @Override
+  public ReadResponseStats getStats() {
+    throw new IllegalStateException(this.getClass().getSimpleName() + " does not support accumulating metrics.");
+  }
+
+  @Override
+  public ReadResponseStatsRecorder getStatsRecorder() {
+    return new CompositeReadResponseStatsRecorder(this.chunks);
+  }
+
+  @Override
+  public ByteBuf getResponseBody() {
+    ByteBuf[] byteBufChunks = new ByteBuf[chunks.length];
+    for (int i = 0; i < chunks.length; i++) {
+      byteBufChunks[i] = this.chunks[i].getResponseBody();
+    }
+    return Unpooled.wrappedBuffer(byteBufChunks);
+  }
+
+  @Override
+  public int getResponseSchemaIdHeader() {
+    return this.chunks[0].getResponseSchemaIdHeader();
+  }
+
+  @Override
+  public String toString() {
+    return this.getClass().getSimpleName() + "(" + this.chunks.length + " " + this.chunks[0].getClass().getSimpleName()
+        + " chunks)";
+  }
+
+  private static final class CompositeReadResponseStatsRecorder implements ReadResponseStatsRecorder {
+    /** The aggregated stats of all the chunks (for stats which can be aggregated). */
+    private final ReadResponseStatsRecorder mergedStats;
+
+    /**
+     * This array references the {@link ReadResponseStatsRecorder} of all response chunks beyond the 1st one (for stats
+     * which cannot be aggregated).
+     */
+    private final ReadResponseStatsRecorder[] statsRecorders;
+
+    CompositeReadResponseStatsRecorder(MultiKeyResponseWrapper[] responseChunks) {
+      this.mergedStats = responseChunks[0].getStatsRecorder();
+
+      /**
+       * This array can be one element shorter than {@param responseChunks} because the first chunk's storage exec sub
+       * wait time will be recorded as part of the {@link mergedStats}.
+       */
+      this.statsRecorders = new ReadResponseStatsRecorder[responseChunks.length - 1];
+      ReadResponseStatsRecorder statsRecorder;
+      for (int i = 1; i < responseChunks.length; i++) {
+        statsRecorder = responseChunks[i].getStatsRecorder();
+        // We merge the stats of all chunks from the 2nd one to the last one into the stats of the 1st chunk.
+        this.mergedStats.merge(statsRecorder);
+
+        /**
+         * We also keep a handle to each {@link ReadResponseStatsRecorder} beyond the 1st so that later on we can call
+         * {@link #recordUnmergedMetrics(ServerHttpRequestStats)} on them. */
+        this.statsRecorders[i - 1] = statsRecorder;
+      }
+    }
+
+    @Override
+    public void recordMetrics(ServerHttpRequestStats stats) {
+      this.mergedStats.recordMetrics(stats);
+      recordUnmergedMetrics(stats);
+    }
+
+    @Override
+    public void recordUnmergedMetrics(ServerHttpRequestStats stats) {
+      for (int i = 0; i < this.statsRecorders.length; i++) {
+        this.statsRecorders[i].recordUnmergedMetrics(stats);
```

Comment:
Not sure if I missed something but besides preference, any other reason to `recordUnmergedMetrics` for the first chunk as part of `recordMetrics` and have `statsRecorders` array to be off by one (excluding the first chunk)?

Reply:
The way I did it is definitely not the only way...

For example, another way could have been to keep a handle of all the statsRecorders in the array, and call `recordUnmergedMetrics` on all of them. As for the mergeable metrics, I could have created a new (empty) stats recorder and merge into that one. But a difficulty of doing it this way is that I don't necessarily have access to the constructor of the correct stats recorder type (I would have needed to wire it in, or else add one more function to the interface to get an empty instance).

I also had an earlier iteration where I added a getter for the storage execution submission wait time and I gathered an array of just these data points... this might have been slightly more lightweight but I thought the design was too brittle. Adding metrics would require more custom coding, which I thought was already a problem in the legacy code (adding a metric required adding getters and setters into `ReadResponse`, and `StatsContext` and copying each field one by one, and all that...).

So I tried to pick a general approach that would be easy to extend with new metrics. But it's certainly possible that there could be an even better way.

Reply:
I see, thanks for the detailed explanation. Was curious to understand the reasoning behind this specific approach.